File: python/AddBinary.py
class AddBinary(object):
    def addBinary(self,a,b):
        l=[]
        carry=0
        
        if len(a)>=len(b):
            i=len(a)-1
            for m in b[::-1]:
                t=int(a[i])+int(m)+carry
                carry=self.add(t,l)
                i=i-1
        else:
            i=len(b)-1
            for n in a[::-1]:
                t=int(n)+int(b[i])+carry
                carry=self.add(t,l)
                i=i-1
        if len(a)>len(b):
            for n in a[-len(b)-1::-1]:
                t=int(n)+carry
                carry=self.add(t,l)
        if len(a)<len(b):
            for m in b[-len(a)-1::-1]:
                t=int(m)+carry
                carry=self.add(t,l)
        if carry==1:
            l.insert(len(l),'1')
            print l
        l.reverse()
        return ''.join(l)
    def add(self,t,l):
        if t==3:
            l.append('1')
            carry=1
        elif t==2:
            l.append('0')
            carry=1
        elif t==1:
            l.append('1')
            carry=0
        else:
            l.append('0')
            carry=0
        return carry


addb=AddBinary()
addb.addBinary("1","1")

File: C/AddBinary.c
#include <stdlib.h>
#include <string.h>

#define max(a, b) (((a) > (b)) ? (a) : (b))

void reverse(char* arr, int first, int last)
{
	while (first < last)
	{
		char tmp = arr[first];
		arr[first] = arr[last];
		arr[last] = tmp;
		++first;
		--last;
	}
}

char* addBinary(char* a, char* b)
{
	int i = strlen(a), j = strlen(b), resSize = 0, carry = 0;
	char* res = (char*)malloc(sizeof(char) * (max(i, j) + 2));
	for (--i, --j; i >= 0 || j >= 0 || carry > 0;)
	{
		if (i >= 0)
			carry += a[i--] - '0';
		if (j >= 0)
			carry += b[j--] - '0';
		res[resSize++] = carry % 2 + '0';
		carry /= 2;
	}
	reverse(res, 0, resSize - 1);
	res[resSize] = '\0';
	return res;
}
